drivers: serial: stm32: use hardware rx timeout interrupt when available

kamnxt opened the pull request:
Use the RTO interrupt for RX timeouts when using the async API. The workqueue based approach is sensitive to load and has a higher latency than using hardware timeouts.

Enable when the register is defined and we're on a normal (not LP) uart.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -59,6 +59,15 @@ LOG_MODULE_REGISTER(uart_stm32, CONFIG_UART_LOG_LEVEL);
+/* Helper for checking if we can use hardware receive timeouts
+ * instead of the work queue
+ */
+#ifdef CONFIG_SOC_SERIES_STM32U5X
```
I'm not quite fond of using series Kconfig directly in files.
I'd prefer you add a Kconfig for the feature and protect it this way, then series that support this feature can be added directly in the Kconfig file rather than having a `#idef` OR'ed list of series poping up in the driver code.

kamnxt replied:
Changed the check now. Hopefully this is a good way of doing it?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1369,7 +1378,16 @@ static void uart_stm32_isr(const struct device *dev)
-	if (LL_USART_IsEnabledIT_IDLE(usart) &&
+
+	if (HAS_RTO(config->usart) && LL_USART_IsEnabledIT_RTO(usart) &&
+	    LL_USART_IsActiveFlag_RTO(usart)) {
```
This won't compile for series not supporting RTO (L1/F1 from what I can see quickly)

erwango reviewed (changes requested):
Thanks for working on this.
To avoid having the feature unevenly supported for years across series that have required HW, I'd like to enable it directly to the maximum extent (and avoid 20 PR to extend it series by series).
Is there a sample that allow to test it in tree ?
If yes, we can provide the support to test it widely on compatible HW.

kamnxt commented on drivers/serial/uart_stm32.c:
```diff
@@ -1369,6 +1374,14 @@ static void uart_stm32_isr(const struct device *dev)
+			LL_USART_ClearFlag_RTO(usart);
+			uart_stm32_dma_rx_flush(dev, STM32_ASYNC_STATUS_TIMEOUT);
+		}
+	}
```
Ended up shuffling this around a bit to get the include guards a bit cleaner

mkaranki commented on drivers/serial/uart_stm32.c:
```diff
@@ -66,6 +66,20 @@ LOG_MODULE_REGISTER(uart_stm32, CONFIG_UART_LOG_LEVEL);
 #define HAS_DRIVER_ENABLE 0
 #endif
 
+/* Helper for checking if we can use hardware receive timeouts
```
Just commenting without know much any real facts.

Would RTO be expected to work with system power management enabled, when MCU enters into stop mode where it basically waits either rx timeout or reception of data?

According to stm32u5 programming manual, RTO interrupt does not exit from stop modes so the receiver timeout never happens. On the other hand, delayed work / workqueue is timer based, thus it would exit the stop mode.

So would the driver need to fall-back into using delayed work when system power management is in use?

kamnxt commented on drivers/serial/uart_stm32.c:
```diff
@@ -66,6 +66,20 @@ LOG_MODULE_REGISTER(uart_stm32, CONFIG_UART_LOG_LEVEL);
+ */
+#ifdef USART_RTOR_RTO
+#define HAS_RTO 1
+#ifdef HAS_LPUART
```
Whoops! Thanks, fixed.

djiatsaf-st commented:
> > @djiatsaf-st Would you be able to verify this PR on whole span of compatible hw?
> 
> Yes, it's possible, I'll run some tests and see what happens.


 Test used :  tests/drivers/uart/uart_async_api/ 

**Build failed:**   

```
zephyrproject/zephyr/drivers/serial/uart_stm32.c: In function 'uart_stm32_isr': 

zephyrproject/zephyr/drivers/serial/uart_stm32.c:75:39: error: implicit declaration of function 'IS_LPUART_INSTANCE'; did you mean 'IS_UART_INSTANCE'? [-Werror=implicit-function-declaration] 

   75 | #define IS_UART_RTO_INSTANCE(usart) (!IS_LPUART_INSTANCE(usart)) 

      |                                       ^~~~~~~~~~~~~~~~~~ 
/local/ztb/zephyrproject/zephyr/drivers/serial/uart_stm32.c:1428:20: note: in expansion of macro 'IS_UART_RTO_INSTANCE' 

1428 |         } else if (IS_UART_RTO_INSTANCE(config->usart) && LL_USART_IsEnabledIT_RTO(usart) && 

      |                    ^~~~~~~~~~~~~~~~~~~~ 
```

**Boards concerned :**  nucleo_c071rb, nucleo_f091rc, stm32f3_disco, nucleo_f746zg(.async_api) 

**Test failed:**  
```
Running TESTSUITE uart_async_timeout 
=================================================================== 
UART instance:serial@40013800 
START - test_forever_timeout 

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:800: uart_async_timeout_test_forever_timeout: (k_sem_take(&rx_rdy, K_MSEC(1000)) equal to 0) 

RX_RDY timeout 


FAIL - test_forever_timeout in 1.020 seconds
``` 

boards concerned : disco_l475_iot1, nucleo_g071rb, nucleo_g474re, nucleo_h743zi, nucleo_wl55jc, stm32n6570_dk//sb , nucleo_f746zg (nocache_mem_dt.nucleo_f746zg, nocache_mem_dt) 



**stm32u083c_dk** has several testcases failed due to timeout 
```
[...]

Running TESTSUITE uart_async_read_abort 
================================================================== 
UART instance:serial@40004800 

START - test_read_abort 
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:625: uart_async_read_abort_test_read_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0) 

RX_RDY timeout 

FAIL - test_read_abort in 0.120 seconds 

=================================================================== 
TESTSUITE uart_async_read_abort failed. 

Running TESTSUITE uart_async_single_read 
=================================================================== 
UART instance:serial@40004800 

START - test_single_read 
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:226: uart_async_single_read_test_single_read: (k_sem_take(&rx_rdy, K_MSEC(105)) not equal to 0) 

RX_RDY timeout 

FAIL - test_single_read in 0.226 seconds 
=================================================================== 
TESTSUITE uart_async_single_read failed. 

[...]

```

**Tests succeeded** on the following series, for example: F1, F2, F4, L1, U3, WB, WBA

etienne-lms commented on drivers/serial/uart_stm32.c:
```diff
@@ -1795,11 +1824,27 @@ static int uart_stm32_async_rx_enable(const struct device *dev,
+#if HAS_RTO
+	if (IS_UART_RTO_INSTANCE(usart)) {
+		/* Use hardware RTO interrupt */
+		uint32_t timeout_in_bits = timeout * data->uart_cfg->baudrate / 1000000;
```
`timeout == -1` stands for no timeout (`SYS_FOREVER_US`). It such a case RTO interrupt should simply not be armed. For simplicity, maybe treat any negative timeout value as an no timeout configuration request.

Another issue here actually. Timeout in is microsecond.  If use uses a timeout of 1 or few second, this 32bit multiplication may overflow before the division. I would suggest to use `uint64_t` for the calculus.

etienne-lms reviewed (commented):
@djiatsaf-st, the build error cases should be fixed by the latest update on the P-R.

Looking at the *test_forever_timeout* test error cases, see my review comment below.

As for the stm32u083c_dk, I can't tell as of now.

etienne-lms reviewed (commented):
LGTM. Successfully tested on a few boards I have here.

Some suggestions on the implementation.

kamnxt commented on drivers/serial/uart_stm32.c:
```diff
@@ -66,6 +66,21 @@ LOG_MODULE_REGISTER(uart_stm32, CONFIG_UART_LOG_LEVEL);
+#ifdef USART_RTOR_RTO
+#define HAS_RTO 1
+#if HAS_LPUART
+#define UART_INSTANCE_HAS_RTO_SET(usart)                                                           \
+	(!IS_LPUART_INSTANCE(usart) && (LL_USART_GetRxTimeout(usart) != 0))
```
I also prefer it that way, but clang-format prefers it there...

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1369,8 +1384,8 @@ static void uart_stm32_isr(const struct device *dev)
 		return;
 	}
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
-	if (LL_USART_IsEnabledIT_IDLE(usart) &&
-			LL_USART_IsActiveFlag_IDLE(usart)) {
+
+	if (LL_USART_IsEnabledIT_IDLE(usart) && LL_USART_IsActiveFlag_IDLE(usart)) {
```
Please don't mix formatting changes with functional changes, but use a dedicated commit.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1738,6 +1768,50 @@ static int uart_stm32_async_tx(const struct device *dev,
+	}
+
+	/* Use hardware RTO interrupt */
+	LOG_DBG("timeout: %d us / %d bits", timeout, (uint32_t)timeout_in_bits);
```
```suggestion
	LOG_DBG("Set RTO timeout: %d us / %d bits", timeout, (uint32_t)timeout_in_bits);
```

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1795,11 +1868,10 @@ static int uart_stm32_async_rx_enable(const struct device *dev,
-	LL_USART_ClearFlag_IDLE(usart);
-	LL_USART_EnableIT_IDLE(usart);
+	/* Set timeout if applicable */
+	if (timeout != SYS_FOREVER_US) {
```
This is a behavior change vs application configuration vs initial implementation and would probably require a dedicated commit.

erwango reviewed (changes requested):
@djiatsaf-st, @etienne-lms  Can you remove the "Blocking: HW Tests" if no issue was found when testing ?

kamnxt committed:
drivers: serial: stm32: fix some formatting

Fix some formatting, because the clang-format compliance test was
complaining.

Signed-off-by: Kamil Krzyżanowski <kamnxt@kamnxt.com>

kamnxt committed:
drivers: serial: stm32: use hardware rx timeout interrupt when available

Use the RTO interrupt for RX timeouts when using the async API.
The workqueue based approach is sensitive to load and has a higher latency
than using hardware timeouts.

Signed-off-by: Kamil Krzyżanowski <kamnxt@kamnxt.com>

etienne-lms commented on drivers/serial/uart_stm32.c:
```diff
@@ -1738,6 +1768,54 @@ static int uart_stm32_async_tx(const struct device *dev,
 }
 
+static void set_timeout_itr(USART_TypeDef *usart, uint32_t baudrate, int32_t timeout)
+{
```
Actaully I swa another issue. Using `UART_INSTANCE_HAS_RTO_SET()` from `uart_stm32_isr()` and `uart_stm32_async_rx_disable()` is misleading since the RTO bit-timeout may be sete but the interrupt not enabled. I think these occurrences should be replaced with a test on the interrupt enable flag (`LL_USART_IsEnabledIT_RTO()`) that is 0 by SoC design when RTO is not supported for the UART instance.

etienne-lms commented on drivers/serial/uart_stm32.c:
```diff
@@ -1736,6 +1767,63 @@ static int uart_stm32_async_tx(const struct device *dev,
+	 * if the timeout requested is too short or 0.
+	 */
+	LL_USART_SetRxTimeout(usart, (uint32_t)timeout_in_bits);
+	if (!UART_INSTANCE_HAS_RTO_SET(usart)) {
```
Nitpicking: now that `UART_INSTANCE_HAS_RTO_SET()` is used only once, here, it may be worth to remote the macro:
```suggestion
	if (LL_USART_GetRxTimeout(usart) != 0)

UART_INSTANCE_HAS_RTO_SET

kamnxt replied:
Good point (but you flipped the logic in your suggestion ;) )

etienne-lms reviewed (approved):
LAGTM. Tests are all ok. Thanks @djiatsaf-st for the help.
